sdk/python/lib/pulumi/provider/experimental/ComponentProvider.md

```diff
@@ -0,0 +1,185 @@
+
+```python
+from typing import Optional
+
+import pulumi
+
+class MyComponent(pulumi.ComponentResource):
+    def __init__(self, name: str, opts: Optional[pulumi.ResourceOptions] = None):
+        super().__init__("my-provider:index:MyComponent", name, {}, opts)
+        self.register_outputs({})
+```
+
+> [!NOTE]
+> The type string `my-provider:index:MyComponent` follows the pattern `<package>:<module>:<type>`. The package is your provider's name, module must be `index`. The `type` must match the class name.
+
+Next, create a `__main__.py` file to host the provider:
+
+```python
+from pulumi.provider.experimental import Metadata, component_provider_host
+
+if __name__ == "__main__":
+    component_provider_host(
+        Metadata(name="my-provider", version="1.2.3", display_name="My Component Provider")
+    )
+```
+
+> [!NOTE]
+> The name passed to `Metadata` must exactly match the package name in the type string of your component.
+
+The call to `component_provider_host` will start the provider and listen for RPC requests from the Pulumi engine. Any subclasses of `pulumi.ComponentResource` found in any Python source code files in the provider's directory will be exposed as resources by the provider.
+
+At this point, we can run the provider, but it does not yet do anything useful. To get data in and out of the provider we need to define the inputs and outputs of the component. Inputs are infered from the `args` parameter of the component's constructor. Add a new type `MyComponentArgs` derived from `TypedDict` to represent the arguments, and add it as an argument to the constructor.
+
+```diff
+-from typing import Optional
++from typing import Optional, TypedDict
+
+ import pulumi
+
++class MyComponentArgs(TypedDict):
++    who: Optional[pulumi.Output[str]]
++
+ class MyComponent(pulumi.ComponentResource):
+-    def __init__(self, name: str, opts: Optional[pulumi.ResourceOptions] = None):
++    def __init__(self, name: str, args: MyComponentArgs, opts: Optional[pulumi.ResourceOptions] = None):
+         super().__init__("my-provider:index:MyComponent", name, {}, opts)
+         self.register_outputs({})
+```
+
+Outputs are defined as attributes on the component class:
+
+```diff
+ class MyComponent(pulumi.ComponentResource):
++    greeting: Optional[pulumi.Output[str]]
++
+     def __init__(self, name: str, args: MyComponentArgs, opts: Optional[pulumi.ResourceOptions] = None):
+         super().__init__("my-provider:index:MyComponent", name, {}, opts)
+         self.register_outputs({})
+```
+
+To return data from the component, assign a value to the output attribute:
+
+```diff
+     def __init__(self, name: str, args: MyComponentArgs, opts: Optional[pulumi.ResourceOptions] = None):
+         super().__init__("my-provider:index:MyComponent", name, {}, opts)
```

Comment:
Should we extend the guide to create at least one resource? Then, also explain the naming and parenting requirements?